### PR TITLE
Create SelectableList kit

### PR DESCRIPTION
**Pull request by josephinef9**
# Screen Shots

Screen Shot 2020-09-14 at 3 09 01 PM

#### Runway Ticket URL

[CMPN-13](https://nitro.powerhrg.com/runway/backlog_items/CMPN-13) - 
Finish ListRadio kit / turn it into the SelectableList kit

# Issues

As per @garettarrowood we are submitting the first part of this story without the radio button and checkbox selected state. We've come to an issue were the radio button highlighting would add too much complexity to this story and the current state adds enough value, CR pending.

## What we tried

We were able to get the React checkbox portion up and running fine, but possibly not how UX envisioned the story. We had to add in state to the `SelectableListItem` component and a custom event listener that returns the `onEvents` prop. The logic for the radio button selection (from our various trials and errors) may need to be in the `_list_item.jsx` file under separate conditional logic. Here's how we think that can be accomplished:

1. Add state to the component that listens for a change in any of the children when the components are the `radio` type
1. Add a callback to `SelectableListItem` that fires when a button is pressed
1. the callback would then get the current selected item's id and apply a selected style over the current item and the default color to all the remaining items

There may be a cleaner way todo that, but we have been told that it is out of scope for this story and should be broken into a separate one that is handled by UX!

#### Checklist:

- [ ] **DEPLOY** Please add the `Milano` label when you are ready for a review
- [x] **SCREENSHOT** Please add a screen shot or two
- [x] **SPECS** Please cover your changes with specs

### CODE REVIEW

File: app/pb_kits/playbook/_playbook.scss [hunk 1]
```diff
@@ -81,3 +81,5 @@
 @import 'pb_time_stacked/time_stacked';
 @import 'pb_weekday_stacked/weekday_stacked';
 @import './utilities/spacing';
+
+@import 'pb_selectable_list/selectable_list';
```
Comment (garettarrowood) [inline app/pb_kits/playbook/_playbook.scss:85]: Be sure to put this in alpha order above

File: app/pb_kits/playbook/data/menu.yml [hunk 2]
```diff
@@ -94,3 +94,4 @@ kits:
     - weekday_stacked
   - user
   - user_badge
+  - selectable_list
```
Comment (garettarrowood) [inline app/pb_kits/playbook/data/menu.yml:97]: Same here - move to alpha order

File: app/pb_kits/playbook/index.js [hunk 3]
```diff
@@ -95,3 +96,6 @@ export PbTypeahead from './pb_typeahead'
 export PbPopover from './pb_popover'
 export PbTable from './pb_table'
 export PbTooltip from './pb_tooltip'
+
+export SelectableList from './pb_selectable_list/_selectable_list.jsx'
+export SelectableListItem from './pb_selectable_list/_selectable_list_item.jsx'
```
Comment (garettarrowood) [inline app/pb_kits/playbook/index.js:101]: Same here.

File: app/pb_kits/playbook/packs/examples.js [hunk 4]
```diff
@@ -173,3 +173,5 @@ WebpackerReact.setup({
   ...WeekdayStacked,
 })
 
+import * as SelectableList from 'pb_selectable_list/docs'
+WebpackerReact.setup(SelectableList)
```
Comment (garettarrowood) [inline app/pb_kits/playbook/packs/examples.js:177]: Same here

File: app/pb_kits/playbook/pb_body/_body.jsx [hunk 5]
```diff
@@ -26,7 +26,7 @@ const Body = (props: BodyProps) => {
     aria = {},
     className,
     children,
-    color = '',
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_body/_body.jsx:29]: Is this change intended?

File: app/pb_kits/playbook/pb_caption/_caption.scss [hunk 6]
```diff
@@ -16,4 +16,7 @@
   &[class*=dark]{
     @include caption_dark
   }
+  &[class*=dark]{
+    @include caption_dark
+  }
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_caption/_caption.scss:21]: Looks like this sneaked in. Looks like a dup of the above and in need of removal.

File: app/pb_kits/playbook/pb_form/docs/_form_simple_form.html.erb [hunk 7]
```diff
@@ -52,7 +52,10 @@
   <%= form.select :example_select_field, [["Yes", 1], ["No", 2]], props: { label: true } %>
   <%= form.collection_select :example_collection_select_field, example_collection, :value, :name, props: { label: true } %>
   <%= form.check_box :example_checkbox_field, props: { text: "Example Checkbox", label: true } %>
+<<<<<<< HEAD
+=======
   <%= form.date_picker :example_date_picker_field1, props: { default_date: 'blank' } %>
+>>>>>>> master
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_form/docs/_form_simple_form.html.erb:58]: Breaking change during merging updated branch.

File: app/pb_kits/playbook/pb_form/docs/_form_simple_form_validate.html.erb [hunk 8]
```diff
@@ -52,7 +52,10 @@
   <%= form.select :example_select_field, [["Yes", 1], ["No", 2]], props: { required: true, blank_selection: "Select One...", label: true } %>
   <%= form.collection_select :example_collection_select_field, example_collection, :value, :name, props: { required: true, blank_selection: "Select One...", label: true } %>
   <%= form.check_box :example_checkbox_field, props: { text: "Example Checkbox", label: true, required: true } %>
+<<<<<<< HEAD
+=======
   <%= form.date_picker :example_date_picker_field2, props: { default_date: 'blank', required: true } %>
+>>>>>>> master
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_form/docs/_form_simple_form_validate.html.erb:58]: Same here.

File: app/pb_kits/playbook/pb_progress_step/progress_step_item.rb [hunk 9]
```diff
@@ -14,6 +14,7 @@ class ProgressStepItem
       def classname
         generate_classname("pb_progress_step_item", status)
       end
+
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_progress_step/progress_step_item.rb:17]: Remove extra line.

File: app/pb_kits/playbook/index.js [hunk 10]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (garettarrowood) [inline app/pb_kits/playbook/index.js:1]: Remove extra line.

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.html.erb [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    data: object.data,
+    id: object.id) do %>
+      <%= pb_rails("list") do %>
+        <% object.items.each do |item| %>
+          <%= pb_rails("selectable_list/selectable_list_item", props: item.merge(variant: object.variant) )%>
+        <% end %>
+      <% end %>
+<% end %>
+
+
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list.html.erb:13]: 1 extra line

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        className={classes}
+        id={id}
+    >
+      <List {...props}>
+        {/* <If condition={props.variant === 'checkbox'}> */}
+        {children}
+        {/* <Else /> */}
+        {/* add in state logic to manage highlight here */}
+        {/* {children} */}
+        {/* </If> */}
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx:43]: Remove commented out code.

File: app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.jsx [hunk 13]
```diff
@@ -0,0 +1,32 @@
+import React from 'react'
+import { SelectableList, SelectableListItem } from '../..'
+
+const SelectableListDefault = () => {
+  return (
+
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.jsx:6]: extra line

File: app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_radio.jsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import React from 'react'
+import { SelectableList, SelectableListItem } from '../..'
+
+const SelectableListDefault = () => {
+  return (
+
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_radio.jsx:6]: extra line

File: app/pb_kits/playbook/pb_selectable_list/selectable_list_item.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+      prop :value, type: Playbook::Props::String,
+                   default: "radio_text"
+
+      prop :variant, type: Playbook::Props::Enum,
+                     values: %w[radio checkbox],
+                     default: "radio"
+
+      partial "pb_selectable_list/selectable_list_item"
+
+      ##todo check this after done everything else
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/selectable_list_item.rb:24]: If this works, remove the todo

File: app/pb_kits/playbook/pb_selectable_list/selectable_list_item.rb [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+      def classname
+        generate_classname("pb_item_kit")
+      end
+
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/selectable_list_item.rb:28]: extra line

File: spec/pb_kits/playbook/kits/body_spec.rb [hunk 17]
```diff
@@ -20,11 +20,11 @@
...
   describe "#classname" do
     it "returns namespaced class name", :aggregate_failures do
-      expect(subject.new({}).classname).to eq "pb_body_kit"
-      expect(subject.new(classname: "additional_class").classname).to eq "pb_body_kit additional_class"
-      expect(subject.new(dark: true).classname).to eq "pb_body_kit dark"
+      expect(subject.new({}).classname).to eq "pb_body_kit_default"
+      expect(subject.new(classname: "additional_class").classname).to eq "pb_body_kit_default additional_class"
+      expect(subject.new(dark: true).classname).to eq "pb_body_kit_default dark"
       expect(subject.new(dark: true, color: "light").classname).to eq "pb_body_kit_light dark"
-      expect(subject.new(status: "negative").classname).to eq "pb_body_kit_negative"
```
Comment (garettarrowood) [inline spec/pb_kits/playbook/kits/body_spec.rb:27]: As far as I can see, none of this should be in here. @thestephenmarshall - any ideas how this change keeps cropping up? I saw something like this in a broken Nitro-Web CI run 2 weeks ago.

**Review by garettarrowood — APPROVED** [review]
Yay!

**Comment by christinaatai** [comment]
There is some extra padding on the react version. Can you remove the extra padding?

Screen Shot 2020-09-15 at 9 22 51 AM

**Comment by christinaatai** [comment]
I have also linked [List Radio Kit](https://github.com/powerhome/playbook/issues/185) and [List Checkbox Kit](https://github.com/powerhome/playbook/issues/184) issues in the PR. In the screenshots in those issues, there is some padding to the left of the radio and checkbox that is missing in your PR. You can do this by adding spacing props to the radio and checkbox kit.

File: app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+          variant: "checkbox",
+          items: [
+            { text: "one", name: "custum1" },
+            { text: "two", name: "custum2" },
+            { text: "three", name: "custum3" },
+            { text: "four", name: "custum4" },
+          ]
+        }
+    )
+%>
```
Comment (christinaatai) [inline app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb:12]: @jasoncypret We usually have the item as their own kit. This works, but it's not consistent with how the other kit works. Should we continue with this?

**Review by christinaatai — CHANGES_REQUESTED** [review]

**Comment by josephinef9** [comment]
> There is some extra padding on the react version. Can you remove the extra padding?
> 
> Screen Shot 2020-09-15 at 9 22 51 AM

We went in and fixed the padding on our last commit. It looks the same on our end. Are the screenshots from today?

File: app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb (hunk 18 above)
Comment (garettarrowood) [reply app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb:12]: We should. We worked hard at making this invoke the child kit. But Ruby won't behave the same way JS/React will here.  @andrew-k9 - This PR was supposed to contain lots of notes about this effort. Can you please elaborate on why we had to go with this design?

Comment (garettarrowood) [reply app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb:12]: Oh wait. It does!! @christinaatai - Check out the "What we tried" section in the summary. :)

Comment (jasoncypret) [reply app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb:12]: Another thing to keep in mind is many of the form related kits have an option to bring your own input. This is because when using these with simple form it's best to just wrap the input in the kit. Just another potential reason to build items as sub kits.

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+  variant?: 'checkbox' | 'radio',
+}
+
+const SelectableList = (props: SelectableListProps) => {
+  const {
+    aria = {},
+    children,
+    className,
+    data = {},
+    id,
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx:23]: `variant` needs to be destructured from the `SelectableListProps`, and defaulted to `radio`

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  const dataProps = buildDataProps(data)
+
+  return (
+    <div
+        {...ariaProps}
+        {...dataProps}
+        className={classes}
+        id={id}
+    >
+      <List {...props}>
```
Comment (garettarrowood) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx:37]: But then still be sure to include it with the spread props here.

**Comment by christinaatai** [comment]
> > There is some extra padding on the react version. Can you remove the extra padding?
> > Screen Shot 2020-09-15 at 9 22 51 AM
> 
> We went in and fixed the padding on our last commit. It looks the same on our end. Are the screenshots from today?

@josephinef9 Yeah, but I don't think it was pulled correctly the first time I ran it today. I see the change now!

File: app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb (hunk 18 above)
Comment (andrew-k9) [reply app/pb_kits/playbook/pb_selectable_list/docs/_selectable_list_checkbox.html.erb:12]: FWIW this is the only way to add `selectable_list_item`s to the `selectable_list`. We tried an approach like 

```ruby
<%= pb_rails("selectable_list", props: { variant: "checkbox" }) do %>
  <%= pb_rails("selectable_list/selectable_list_item", props: { text: "Unselected", name: "custum1" }) %>
        props: {
  <%= pb_rails("selectable_list/selectable_list_item", props: {text: "Selected", name: "custum1"}) %>
          variant: "radio",
  <%= pb_rails("selectable_list/selectable_list_item", props: {text: "Hover", name: "custum1"}) %>
          items: [
  <%= pb_rails("selectable_list/selectable_list_item", props: {text: "Unslected Last", name: "custum1"}) %>
            { text: "one", name: "custum1" },
<% end %>
```

before and it does not allow us to add the `variant` prop to `selectable_list`s children. In 
```ruby
capture(&object.children)
```
we would first need to take each child element, give it a `variant` prop, and then have `capture` present the block to the `selectable_list_item.html.erb` file (unsure of wording for that 🤔) which I don't think is possible with procs in ruby without doing something like 

```ruby
pb_rails("selectable_list") do |variant|
```
when being called 😳

**Comment by josephinef9** [comment]
> I have also linked [List Radio Kit](https://github.com/powerhome/playbook/issues/185) and [List Checkbox Kit](https://github.com/powerhome/playbook/issues/184) issues in the PR. In the screenshots in those issues, there is some padding to the left of the radio and checkbox that is missing in your PR. You can do this by adding spacing props to the radio and checkbox kit.

We're taking a look at the issues page you linked and can't really distinguish what the padding on left is. Can you be a little more specific about how much padding is needed?

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss [hunk 21]
```diff
@@ -0,0 +1,9 @@
+@import "../tokens/colors";
+
+[class^=pb_selectable_list]{
+  .pb_item_kit {
+    &:hover {
+      background-color: rgb(249,251,255);
```
Comment (thestephenmarshall) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss:6]: This is worrysome. We really want to try and use `tokens` if we can for these scenarios: https://playbook.powerapp.cloud/utilities

File: spec/pb_kits/playbook/kits/selectable_list_spec.rb [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+  it { is_expected.to define_enum_prop(:variant)
+                      .with_default("radio")
+                      .with_values("radio", "checkbox")
+  }
+
+  describe "#classname" do
+    it "returns namespaced class name", :aggregate_failures do
+      expect(subject.new({}).classname).to eq "pb_selectable_list_kit"
+      expect(subject.new(classname: "additional_class").classname).to eq "pb_selectable_list_kit additional_class"
+    end
```
Comment (thestephenmarshall) [inline spec/pb_kits/playbook/kits/selectable_list_spec.rb:19]: This is missing a lot of spec coverage.

**Review by thestephenmarshall — CHANGES_REQUESTED** [review]
There are a few things to address here before merging.

**Comment by jasoncypret** [comment]
Marked this as "Future Release" till we test it through an alpha on Nitro

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+  const {
+    aria = {},
+    children,
+    className,
+    data = {},
+    id,
+  } = props
+
+  const ariaProps = buildAriaProps(aria)
+  const classes = classnames(buildCss('pb_selectable_list_kit'), className, globalProps(props))
```
Comment (jasoncypret) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx:27]: className needs to come last so that css modules work correctly in Nitro
```suggestion
  const classes = classnames(buildCss('pb_selectable_list_kit'), globalProps(props), className)
```

Comment (josephinef9) [reply app/pb_kits/playbook/pb_selectable_list/_selectable_list.jsx]: Fixed it

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss (hunk 21 above)
Comment (josephinef9) [reply app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss]: Changed it to use the token

**Comment by christinaatai** [comment]
> > I have also linked [List Radio Kit](https://github.com/powerhome/playbook/issues/185) and [List Checkbox Kit](https://github.com/powerhome/playbook/issues/184) issues in the PR. In the screenshots in those issues, there is some padding to the left of the radio and checkbox that is missing in your PR. You can do this by adding spacing props to the radio and checkbox kit.
> 
> We're taking a look at the issues page you linked and can't really distinguish what the padding on left is. Can you be a little more specific about how much padding is needed?

Let's connect over zoom if this is still unclear.

Artboard

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list_item.html.erb [hunk 24]
```diff
@@ -0,0 +1,18 @@
+<%= content_tag(:li,
+    aria: object.aria,
+    class: object.classname,
+    data: object.data,
+    id: object.id) do %>
+    <% puts object.variant %>
+  <% if object.variant == "radio"%>
+    <%= pb_rails("radio", props: { text: object.text, padding_bottom: "md" } ) %>
```
Comment (christinaatai) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list_item.html.erb:8]: Can you remove padding bottom?

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list_item.html.erb [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+    <% if object.children.present? %>
+      <%= capture(&object.children) %>
+    <% end %>
+  <% else %>
+    <%= pb_rails("checkbox", props: { text: object.text, padding_bottom: "md" } ) %>
```
Comment (christinaatai) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list_item.html.erb:13]: Same here – can you remove the padding here as well?

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss [hunk 26]
```diff
@@ -0,0 +1,9 @@
+@import "../tokens/colors";
+
+[class^=pb_selectable_list]{
+  .pb_item_kit {
+    &:hover {
+      background-color: $bg_light;
+    }
+  }
+}
```
Comment (christinaatai) [inline app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss:9]: The Select Kit (specifically, class `pb_select`)  is overriding the Selectable List Kit in React to give it a margin bottom. Can you add in css so that there is no margin bottom would be applied for the class `pb_selectable_list_item_kit`?

File: spec/pb_kits/playbook/kits/selectable_list_spec.rb (hunk 22 above)
Comment (andrew-k9) [reply spec/pb_kits/playbook/kits/selectable_list_spec.rb:19]: What other test coverage specifically should we add?

File: app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss (hunk 26 above)
Comment (josephinef9) [reply app/pb_kits/playbook/pb_selectable_list/_selectable_list.scss:16]: We went and hopefully fixed the padding issue!

File: spec/pb_kits/playbook/kits/selectable_list_spec.rb (hunk 22 above)
Comment (josephinef9) [reply spec/pb_kits/playbook/kits/selectable_list_spec.rb:24]: @thestephenmarshall We added another line to the selectable_list_spec as well as adding a selectable_list_item_spec file with tests in it. Is there any other test coverage you think we should add?

**Comment by jasoncypret** [comment]
2 things that should have probably have been in the kit spec. 

- [ ] The list item row should be clickable. You could achieve this with CSS only by making the label cover the whole row.
- [ ]  To achieve the selectable highlighting you will like need to add some light JS for rails. For react you will need to handle this with state and onlick (See selectable cards)

The goal of this kit is to use this in several spots. Most commonly will be within a popover or HTML dropdown list. Another use case is on settings pages. Just wanted to fill in some ambiguity.

**Review by garettarrowood — APPROVED** [review]
In regards to what the Champion do, this is complete. NUX - feel free to merge without the 2 items above, or to leave this open and continue to work towards making the rows clickable and highlighted.

**Comment by christinaatai** [comment]
I'm opening a new PR to clean the slate
